Create Doc Parser and appropriate Tests & Benchmarks

BinarySoftware opened the pull request:
### Pull Request Description
This introduces DocParser and DocAST to the Parser. It is going to be used used to generate nicely formatted documentation for Enso code. (#42)

This PR comes with appropriate tests and bench suites.

### Important Notes
- This PR contains changes after CR of first PR, including proper indentation, changes in layout and now it passes tests with assertion of input to output - simply, the parser is now capable of properly showing parsed text, as it is expected from a good parser

### Checklist
Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [ ] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [ ] All code has been tested where possible.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+
+  trait Spanned {
+    def span: Int
+  }
```
The things above should not be implemented here, please import them instead (+ update to the current design)

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  implicit def HasRepr_6[T: ReprOf]: ReprOf[List[T]] =
+    _.foldLeft(Repr.Empty(): Repr)((a, b) => Repr.Seq(a, reprOf(b)))
+  implicit def HasRepr_7[T: ReprOf]: ReprOf[NonEmptyList[T]] =
+    _.foldLeft(Repr.Empty(): Repr)((a, b) => Repr.Seq(a, reprOf(b)))
```
The same applies here, please import `Repr` like things

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+      htmlRepr.show(htmlBuilder)
+      htmlBuilder.result()
+    }
+  }
```
You should also import Symbol from Parser, you can extend it with your own things here, but lets base on parser's one.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+
+  final case class Text(text: String) extends AST {
+    val textRepr: Repr = text
+    val htmlRepr: Repr = text.replaceAll("\n", "<br />")
```
I would rather keep the same name everywhere - `repr` instead of `textRepr`. Regarding HTML, it sohuld nto have a type of `Repr`, you should build HTML using some HTM<L structures that have encoded pretty prinitng option. Each node in HTML should be a structure. Don't implement it by your own, just use a popular library that provides this structures + pretty printing.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+
+  case object Bold extends FormatterType {
+    val showableMarker: Char = '*'
+    val HTMLMarker: Char     = 'b'
```
dont use capitalized names for variables

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  case object Strikethrough extends FormatterType {
+    val showableMarker: Char = '~'
+    val HTMLMarker: Char     = 's'
+  }
```
I would make an abstract class for that, like 
```
abstract class FormatterClass(val showableMarker: Char, htmlMarker: Char) extends FormatterType

case object Bold extends FormatterClass('*','b')
case object Italic extends FormatterClass('_','i')
```

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+    val textRepr
+      : Repr = Repr(formatterType.showableMarker) + elem.textRepr + formatterType.showableMarker
+    val htmlRepr
+      : Repr = Repr("<") + formatterType.HTMLMarker + ">" + elem.htmlRepr + "</" + formatterType.HTMLMarker + ">"
```
Apply code formatting style pls!

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+    val HTMLMarker: Char     = 's'
+  }
+
+  final case class Formatter(formatterType: FormatterType, elem: Option[AST])
```
these names are too long. You are in class `Formatter`, so its field should be named `type`, not `formatterType`. Of course this is a special case where its a keyword, so use `tp` instead. Then you know its `Formatter.tp`.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  /// Unclosed Formatter ///
+  //////////////////////////
+
+  final case class UnclosedFormatter(
```
If this is part of formatter, iut sohuld be in its namespace, like `Formatter.Unclosed`. Apply in other places as well.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  final case class CodeLine(code: String, inMultilineCode: Boolean)
+      extends AST {
+    val textRepr: Repr = {
+      if (inMultilineCode) {
```
This is bad – AST element should not know about its context. You should not have a field `inMultilineCode` flag. Make either separate datatype here.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  ////// Header //////
+  ////////////////////
+
+  final case class Header(elem: AST) extends AST {
```
As written above, such thigns as `Header` sohuld be in their namespace. I understand that this is something like `Block.Header` or something.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+      case Image =>
+        Repr("<img src=\"") + url + "\" alt=\"" + name + "\">"
+    }
+  }
```
This is OK design, but its harder to pattern match on. I would rewrite it to 2 separate case classes `Link.URL` and `Link.Image` (with a base abstract class / trait)

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  ///////////////////
+
+  trait LinkType {
+    val readableMarker: String
```
again, I would change the name to `marker` or `symbol` here

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+    val textRepr: Repr = {
+      var _repr = Repr()
+      elems.toList.foreach(elem => {
+        if (elem.show().contains(" " * (indent + 2))) {
```
This is slow, you should never inspect generated text during text generation.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+        if (elem.show().contains(" " * (indent + 2))) {
+          _repr += '\n'
+          _repr += elem.textRepr
+        } else if (elem.isInstanceOf[InvalidIndent]) {
```
dont use `isInstanceOf`, use pattern matching instead

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+        }
+      })
+      _repr
+    }
```
This code is hard and long. Each element sohuld know how it should print ITSELF, so this code should iterate over elements and ask them to print themselves instead.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+      elems.toList.foreach(elem => {
+        _repr += "<li>"
+        _repr += elem.htmlRepr
+        _repr += "</li>"
```
Again, HTML construction by hand is error prone. You can forget to close a node etc. Lets use library for that.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+    def apply(indent: Int, listType: ListType, elem: AST): ListBlock =
+      ListBlock(indent, listType, NonEmptyList(elem, Nil))
+    def apply(indent: Int, listType: ListType, elems: AST*): ListBlock =
+      ListBlock(indent, listType, NonEmptyList(elems.head, elems.tail.to[List]))
```
Use `org.enso.data.List1` instead of `NonEmptyList`

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+        }
+        _repr += elems(i).textRepr
+      }
+      _repr
```
First of all, this code is almost the same as code above - duplication of code is disallowed. Plus apply things in previous comment.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+
+  ///// Important /////
+  final case class Important(indent: Int, elems: List[AST]) extends Section {
+    override def readableMarker = Some('!')
```
Why its def, not a val?

BinarySoftware replied:
because only then I don't need to override repr in every section, if it is val repr uses marker from Section and not from ex. Important even though I've overridden it

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  abstract class Section extends Symbol {
+    val indent: Int
+    val elems: List[AST]
+    def readableMarker: Option[Char] = None
```
What exactly is `Section`? What are sections that does not have `readableMarker`?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+
+  ///// Info /////
+  final case class Info(indent: Int, elems: List[AST]) extends Section {
+    override def readableMarker = Some('?')
```
When it could happen that I pattern match on `Section.Info` and I read its `readableMarker` and get `None`? If it could not happen, then the implementation is misleading.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+    def apply(indent: Int, elem: AST): Example =
+      Example(indent, elem :: Nil)
+    def apply(indent: Int, elems: AST*): Example =
+      Example(indent, elems.to[List])
```
All the smart constructors for sections are exactly the same. We should refactor it to one structure parametrized by a type of the section.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  ///// Multiline Code /////
+  final case class MultilineCode(indent: Int, elems: List[AST])
+      extends Section {
+    override def readableMarker = Some(' ')
```
This is not a marker! We should not fake markers by whitespaces ;)

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  ////// Body //////
+  //////////////////
+
+  final case class Body(elems: List[Section]) extends AST {
```
What happens when Body has no sections? Is it valid? It applies to all other similar places as well.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+      : Repr = Repr("<div class=\"") + name + "\">" + name + ' ' + version.htmlRepr + "</div>"
+  }
+
+  case class Deprecated(version: Option[String]) extends TagType
```
keeping version in string is a bad idea :)

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+  object Upcoming {
+    def apply():                Upcoming = Upcoming(None)
+    def apply(version: String): Upcoming = Upcoming(Option(version))
+  }
```
This is just a lot of copy pasted code. Why dont we create an API like `Tag(Modified, Some(version))` instead, and create smart constructors for `Tag` only?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+    def apply(version: String): Upcoming = Upcoming(Option(version))
+  }
+
+  final case class Tags(indent: Int, elems: List[TagType]) extends AST {
```
If I create an AST and set `elems` to `List()`, then I print it to text and I read it back, will I get back a strcuture containing `Tags(..., List())` ?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,572 @@
+
+      if (synopsis != Synopsis()) {
+        _repr += synopsis.textRepr
+      }
```
Parent class sohuld not handle its children. empty synopsis should know how to print itself. Such ifs are wrong here.

BinarySoftware replied:
fixed with .exists()

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,458 @@
+    val html: Seq[Modifier] =
+      Seq(
+        div(`class` := this.getClass.getSimpleName)(
+          for (elem <- elems) yield elem.html
```
just remove every yield in the code.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,458 @@
+  ////// Details //////
+  /////////////////////
+
+  final case class Details(elems: List[Section]) extends AST {
```
remember to change names in such a way they make sense. I dont get what is `DocAST.Details`

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,458 @@
+      if (details.repr == Repr()) {
+        Repr(name)
+      } else {
+        Repr(name) + ' ' + details.repr
```
we dont ue hardcoded literals in code.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,458 @@
+  case object Upcoming   extends TagType
+
+  final case class TagClass(tp: TagType, details: Option[String]) extends AST {
+    val name: String = tp.getClass.getSimpleName.toUpperCase.dropRight(1)
```
magic number. Why `1` and not `78`?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,458 @@
+        )(if (synopsis.exists()) synopsis.html else "".html)(
+          if (body.exists()) body.html else "".html
+        )
+      )
```
refactor to vriables.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,458 @@
+      )
+  }
+
+  object Documentation {
```
I would rename `Documentation` to just `Doc`. Everyone understand it, and its easier to read.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    val cssLink: HTMLTag = link(rel := "stylesheet")(href := "style.css")
+
+    def renderHTML(): HTMLTag =
+      scalatags.Text.all.html(head(docMeta, cssLink), body(html))
```
I would definitely change the `scalatags.Text.all` to `HTML`.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    def renderHTML(): HTMLTag =
+      scalatags.Text.all.html(head(docMeta, cssLink), body(html))
+
+    def makeIndent(size: Int) = " " * size
```
Why it is a method on `Symbol`? What is the use case when you will have an AST node and you will call `node.makeIndent(5)`?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    final case class Unclosed(tp: Type, elem: Option[AST]) extends InvalidAST {
+      val repr: Repr = Repr(tp.marker) + elem.repr
+      val html: HTML = {
+        val className = s"unclosed_${tp.htmlMarker.tag}"
```
Generating css classes for the elements is really great idea. Please think how you can generalize it, so for all other types they will be generated automatically. Of course sometimes, like here, you would liek to override the automatic generation (`unclosed`) with something custom. Here for example, I'd vote for applying multiple CSS classes - both `.bold` as well as `.unclosed` - this way you can write generic CSS rules.

wdanilo replied:
However, we should not do it in this PR. It gets too long aalready. Let's leave fixing HTML / CSS generation for another task after this will be closed. Please add comment to the code with a note about it.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+  ////// Code Line //////
+  ///////////////////////
+
+  final case class InlineCode(str: String) extends AST {
```
After thinking for a while about it, I would change the name of AST here to something else. AST does not describe well enough such items as InlineCode. In fact your AST has only 1 habitant - `Doc`. Everything else are other strucutres which build up the `Doc`. Think about better name here. Moreover. I believe that `InlineCode`, `Link`, etc, should be grouped in a namespace.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+  ////// Lists //////
+  ///////////////////
+
+  final case class ListBlock(indent: Int, tp: ListBlock.Type, elems: List1[AST])
```
Why not `Block.List` ?

BinarySoftware replied:
Because it should be simply List, but it is impossible

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    final case object Ordered extends Type {
+      val marker     = '*'
+      val HTMLMarker = ol
+    }
```
Repetitive code.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    /// Invalid Indent ///
+    //////////////////////
+
+    final case class InvalidIndent(
```
-> `Indent.Invalid`

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+      indent: Int,
+      elem: AST,
+      tp: Type
+    ) extends InvalidAST {
```
-> `AST.Invalid`

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+
+      _repr += (indent match {
+        case 0 => markerEmptyInNull
+        case 1 => markerNonEmptyInNull
```
I dont understand what this code does. Could you please explain the logic here?

BinarySoftware replied:
So there are some cases in which you want marker to be displayed in the first line. If there is no indent, there should be only marker, or if Null then "", if there is one indent there should be one " " plus marker, but if there is marker:Null, then there should be only " ", not "  "(two spaces) - this pattern match is used to properly indent if there is marker or no - in case of Raw text,

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+      _repr += (indent match {
+        case 0 => markerEmptyInNull
+        case 1 => markerNonEmptyInNull
+        case _ => makeIndent(indent - 2) + markerNonEmptyInNull + makeIndent(1)
```
literals in code. `-2` should never appear in code without proper explanation where this number cames from. Expalantion = good var name. The same applies to `1` **and any literal in the code**

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+      val marker: Option[Char] = Some('>')
+    }
+    case object Code extends Type {
+      val marker: Option[Char] = None
```
repetitive code

wdanilo replied:
Please make sure that in the next PR no code will be repetitive.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+  ////// Synopsis //////
+  //////////////////////
+
+  final case class Synopsis(elems: List[Section]) extends AST {
```
this type is wrong. It allows creating synopsis next to bold, or inline code. Make sure the types does not allow for cosntructing wong AST! Please make sure that all other places are fixed as well. This is the most important design part.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    def apply(tp: Type, details: String): Tag =
+      Tag(tp, Option(details))
+
+    trait Type
```
why not sealed?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+      details.repr match {
+        case Repr.R => Repr(name)
+        case _      => Repr(name) + details.repr
+      }
```
I dont get the reason of this code. Could you explain it pls? :)

BinarySoftware replied:
If details.repr is empty then it displays only representation of Name, else displays tag Name with its details, like version or why something has been deprecated

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    details: Details
+  ) extends AST {
+    val repr: Repr = {
+      var _repr = Repr()
```
dont use `var`

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+    val html: HTML = {
+      val tagsHtml     = if (tags.exists()) tags.html else "".html
+      val synopsisHtml = if (synopsis.exists()) synopsis.html else "".html
+      val detailsHtml  = if (details.exists()) details.html else "".html
```
parent object should not inspect children when printing out.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+  ////// Doc /////
+  ////////////////
+
+  final case class Doc(
```
This should be extracted to top-level and the file should be renamed to `org.enso..... .ast.Doc`. Then the `object AST` will be named `Doc` as well , and everything will be so much nicer, like `Doc.Tag`

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/DocAST.scala:
```diff
@@ -0,0 +1,429 @@
+        _repr += details.repr
+      }
+      _repr
+    }
```
I really dont like this code. Again, parent nodes should not inspect children state to print them,. Every child should be able to print itself. Think if and how you can change that. If you came to a conclusion that its impossible, please describe shortly why.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+      HTML.div(HTML.`class` := className)(tags.html)(synopsis.html)(
+        details.html
+      )
+    )
```
Let's refactor such long multilin e expressions always to vriables. Its hard to read expresions which are 5 lines long and consist of mulitple parentheses across these lines. A much better code would be:
```scala
    val htmlCls = HTML.`class` := className
    val html    = HTML.div(htmlCls)(tags.html)(synopsis.html)(details.html)
    Seq(html)
```

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+  def apply(synopsis: Synopsis, body: Body): Doc =
+    new Doc(Tags(), synopsis, body)
+  def apply(synopsis: Synopsis): Doc =
+    new Doc(Tags(), synopsis, Body(Nil))
```
Why don't we have `Body()` constructor?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+
+  trait Indentable {
+    def makeIndent(size: Int): String = " " * size
+  }
```
TBH I dont like this idea. Traits should be "interfaces" describing what an object could do when inheriting it. Making indent is not a property of an object, because it always throws away every property of the object. The function `makeIndent` just dont suit trait pattern, it should be just a free function in the top level object here.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+
+import org.enso.data.List1
+import scalatags.Text.all._
+import scalatags.Text.{all => HTML, _}
```
This does not match our styling guide, you are importing two times the same function, and also, as I already described in my previous PR, we should not pollute the namespace so much and import everything from some scope. It just makes the code harder to read.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+        HTML.charset := "UTF-8"
+      )
+    val cssLink: HTMLTag =
+      HTML.link(HTML.rel := "stylesheet")(HTML.href := "style.css")
```
What is this for? Why every symbol has now `cssLink`? What is for example `boldedElement.cssLink` usable for?

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+            Repr('\n') + elem.repr
+          case elem =>
+            if (elems.head != elem) {
+              Repr('\n') + makeIndent(indent) + tp.marker + elem.repr
```
We use a lot of literals in the code. Let's refactor them out. A . variable like `newline = Repr('\n')` would be much better.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+          case elem @ (_: ListBlock) => elem.html
+          case elem                  => Seq(HTML.li(elem.html))
+        })
+      }))
```
Please lint the code. There are unnecesary parentheses here.

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+  final case class Body(elems: List[Section]) {
+    val repr: Repr =
+      if (elems == Nil) Repr()
+      else Repr("\n") + elems.map(_.repr.show()).mkString("\n")
```
Why not shorter?
```scala
val head: Repr = if (elems == Nil) R else newline
val repr: Repr = head + elems.map(_.repr.show()).mkString("\n")
```

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+          HTML.div(HTML.`class` := this.getClass.getSimpleName)(
+            elems.map(_.html)
+          )
+        )
```
Too long expr

wdanilo commented on syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala:
```diff
@@ -0,0 +1,405 @@
+
+  final case class Synopsis(elems: List[Section]) {
+    val repr: Repr =
+      if (elems == Nil) Repr() else elems.map(_.repr.show()).mkString("\n")
```
What is this if for?

wdanilo replied:
Fix in other palces as well pls

BinarySoftware commented:
@wdanilo Here's a small list of changes applied today, many of which were necessary after yesterday's talk. What's really important - I didn't get any breaking errors like yesterday.
- Fixed Newline before EOF bug
- Added indent support for single tag, removed from Tags()
- Fixed indents in many places
- removed Text("\n") which was used to indicate end of line, now replaced by looping through AST for creation of single Line(elems: List[AST]) object, which is used to store single line of text
- fixed tests, to work after applied changes
- changes to DocAST - simplified Repr, changed to " R + Foo + Bar +...", using Option in Doc - for example now if body is empty you'll get None instead of Body(List()), more apply() methods
- more minor fixes & changes
